File: src/com/easy/FindModeInBinarySearchTree_501.java
/**
 * 
 */
package com.easy;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class FindModeInBinarySearchTree_501 {
    // 2019年7月24日
    // Use HashMap
    // Time Complexity: O(n)
    // Space Complexity:O(n)
    // Runtime: 8 ms, faster than 23.74%
    // Memory Usage: 37.8 MB, less than 99.26%
    private Map<Integer, Integer> mMap = new HashMap<>();
    private int mMax = 0;
    private List<Integer> mList = new ArrayList<>();

    public int[] findMode(TreeNode root) {
        inorder(root);

        for (int key : mMap.keySet()) {
            if (mMap.get(key) == mMax) {
                mList.add(key);
            }
        }

        int[] mResult = new int[mList.size()];
        int mIndex = 0;
        for (int num : mList) {
            mResult[mIndex++] = num;
        }

        return mResult;
    }

    private void inorder(TreeNode root) {
        if (root == null) {
            return;
        }
        inorder(root.left);
        visit(root.val);
        inorder(root.right);
    }

    private void visit(int mValue) {
        mMap.put(mValue, mMap.getOrDefault(mValue, 0) + 1);
        mMax = Math.max(mMap.get(mValue), mMax);
    }

    // 2019年7月24日
    // Use count via inorder traversal
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 1 ms, faster than 97.26%
    // Memory Usage: 37.6 MB, less than 100.00%
    private int mCount = 0;
    private int mMaxCount = -1;
    private TreeNode mPre;
    private List<Integer> mModes = new ArrayList<>();

    public int[] findMode2(TreeNode root) {
        if (root == null) {
            return new int[0];
        }
        mPre = root;
        inorder2(root);

        int[] mResult = new int[mModes.size()];
        for (int i = 0; i < mModes.size(); i++) {
            mResult[i] = mModes.get(i);
        }

        return mResult;
    }

    private void inorder2(TreeNode root) {
        if (root == null) {
            return;
        }
        inorder2(root.left);
        mCount = (mPre.val == root.val) ? mCount + 1 : 1;

        if (mCount == mMaxCount) {
            mModes.add(root.val);
        } else if (mCount > mMaxCount) {
            mModes.clear();
            mModes.add(root.val);
            mMaxCount = mCount;
        }

        mPre = root;
        inorder2(root.right);
    }
}

File: Golang_src/easy/findModeInBinarySearchTree_501.go
package easy

// Use inorder:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 12 ms, faster than 51.06%
// Memory Usage: 7.6 MB, less than 6.38%
func findMode(root *TreeNode) []int {

	if root == nil {
		return nil
	}

	order := inorder(root)

	if len(order) == 1 {
		return []int{order[0]}
	}

	result := []int{}

	cur := 1
	max := 1

	for i := 1; i < len(order); i++ {
		if order[i] == order[i-1] {
			cur++
		} else {
			if cur == max {
				result = append(result, order[i-1])
			} else if cur > max {
				result = []int{order[i-1]}
				max = cur
			}
			cur = 1
		}
	}

	if cur == max {
		result = append(result, order[len(order)-1])
	} else if cur > max {
		result = []int{order[len(order)-1]}
	}

	return result
}

func inorder3(root *TreeNode) []int {
	if root == nil {
		return nil
	}

	order := []int{}
	order = append(order, inorder3(root.Left)...)
	order = append(order, root.Val)
	order = append(order, inorder3(root.Right)...)

	return order
}

// Use hashmap:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 12 ms, faster than 51.06%
// Memory Usage: 6 MB, less than 53.19%
func findMode2(root *TreeNode) []int {
	maxCount := 0
	hashmap := map[int]int{}

	result := []int{}

	inorder2(root, hashmap, &maxCount)

	for k, v := range hashmap {
		if v == maxCount {
			result = append(result, k)
		}
	}

	return result
}

func inorder2(root *TreeNode, hashmap map[int]int, maxCount *int) {

	if root == nil {
		return
	}

	hashmap[root.Val]++
	*maxCount = max(*maxCount, hashmap[root.Val])

	inorder2(root.Left, hashmap, maxCount)
	inorder2(root.Right, hashmap, maxCount)

}

// Use dfs:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 8 ms, faster than 86.36%
// Memory Usage: 6.02 MB, less than 100.00%
func findMode3(root *TreeNode) []int {
	base, cnt, maxCnt := 0, 0, 0

	result := []int{}

	update := func(x int) {
		if x == base {
			cnt++
		} else {
			base = x
			cnt = 1
		}

		if cnt == maxCnt {
			result = append(result, base)
		} else if cnt > maxCnt {
			maxCnt = cnt
			result = []int{base}
		}
	}

	var inorder func(node *TreeNode)
	inorder = func(node *TreeNode) {
		if node == nil {
			return
		}
		inorder(node.Left)
		update(node.Val)
		inorder(node.Right)
	}

	inorder(root)

	return result
}
